File: backend/ingestion/embedder.py
import json
import logging
import os
import time
import urllib.error
import urllib.request

from tenacity import retry, stop_after_attempt, wait_exponential

logger = logging.getLogger(__name__)

EMBEDDING_MODEL = "jina-embeddings-v3"
EMBEDDING_DIMENSIONS = 1024
EMBEDDING_BATCH_SIZE = 32
EMBEDDING_API_URL = "https://api.jina.ai/v1/embeddings"


def _get_api_key() -> str:
    api_key = os.environ.get("JINA_API_KEY", "").strip()
    if not api_key:
        raise RuntimeError("JINA_API_KEY is not configured")
    return api_key


def _embed_texts_once(texts: list[str], task: str) -> list[list[float]]:
    payload = json.dumps(
        {
            "model": EMBEDDING_MODEL,
            "task": task,
            "input": texts,
        }
    ).encode("utf-8")
    request = urllib.request.Request(
        EMBEDDING_API_URL,
        data=payload,
        headers={
            "Authorization": f"Bearer {_get_api_key()}",
            "Content-Type": "application/json",
        },
        method="POST",
    )

    try:
        with urllib.request.urlopen(request, timeout=120) as response:
            body = json.loads(response.read().decode("utf-8"))
    except urllib.error.HTTPError as exc:
        error_body = exc.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"Jina embedding failed ({exc.code}): {error_body}") from exc

    data = body.get("data") or []
    embeddings = [item.get("embedding") or [] for item in data]
    if len(embeddings) != len(texts) or any(not vector for vector in embeddings):
        raise RuntimeError("Jina returned an incomplete embedding batch")

    return [[float(value) for value in vector] for vector in embeddings]


@retry(wait=wait_exponential(multiplier=1, min=2, max=60), stop=stop_after_attempt(5))
def _embed_texts(texts: list[str], task: str) -> list[list[float]]:
    try:
        return _embed_texts_once(texts, task)
    except RuntimeError as exc:
        message = str(exc).lower()
        if "429" in message or "503" in message:
            time.sleep(2)
            return _embed_texts_once(texts, task)
        raise


def generate_embeddings(batch_texts: list[str]) -> list[list[float]]:
    embeddings: list[list[float]] = []

    for index in range(0, len(batch_texts), EMBEDDING_BATCH_SIZE):
        batch = batch_texts[index : index + EMBEDDING_BATCH_SIZE]
        embeddings.extend(_embed_texts(batch, "retrieval.passage"))

    return embeddings


def embed_query(search_query: str) -> list[float]:
    return _embed_texts([search_query], "retrieval.query")[0]

File: frontend/src/lib/rag/ingestion/embedder.ts
import {
  EMBEDDING_API_URL,
  EMBEDDING_BATCH_SIZE,
  EMBEDDING_MODEL,
} from "../constants";

type JinaEmbeddingResponse = {
  data?: Array<{ embedding?: number[] }>;
  detail?: string;
};

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function getApiKey(): string {
  const apiKey = process.env.JINA_API_KEY?.trim();
  if (!apiKey) {
    throw new Error("JINA_API_KEY is not configured");
  }
  return apiKey;
}

async function embedTexts(
  texts: string[],
  task: "retrieval.query" | "retrieval.passage",
): Promise<number[][]> {
  let lastError: unknown;

  for (let attempt = 0; attempt < 5; attempt += 1) {
    try {
      const response = await fetch(EMBEDDING_API_URL, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${getApiKey()}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          model: EMBEDDING_MODEL,
          task,
          input: texts,
        }),
      });

      const body = (await response.json()) as JinaEmbeddingResponse;

      if (!response.ok) {
        const message =
          body.detail ||
          `Jina embedding failed (${response.status})`;
        const error = new Error(message);

        if ((response.status === 429 || response.status === 503) && attempt < 4) {
          lastError = error;
          await sleep(2 ** attempt * 1000);
          continue;
        }

        throw error;
      }

      const embeddings = body.data?.map((item) => item.embedding ?? []) ?? [];
      if (embeddings.length !== texts.length || embeddings.some((v) => !v.length)) {
        throw new Error("Jina returned an incomplete embedding batch");
      }

      return embeddings;
    } catch (error) {
      lastError = error;
      const message = error instanceof Error ? error.message : String(error);
      if (
        (message.includes("429") || message.includes("503")) &&
        attempt < 4
      ) {
        await sleep(2 ** attempt * 1000);
        continue;
      }
      break;
    }
  }

  throw lastError instanceof Error
    ? lastError
    : new Error("Failed to generate embeddings");
}

export async function generateEmbeddings(batchTexts: string[]): Promise<number[][]> {
  const embeddings: number[][] = [];

  for (let i = 0; i < batchTexts.length; i += EMBEDDING_BATCH_SIZE) {
    const batch = batchTexts.slice(i, i + EMBEDDING_BATCH_SIZE);
    const batchEmbeddings = await embedTexts(batch, "retrieval.passage");
    embeddings.push(...batchEmbeddings);
  }

  return embeddings;
}

export async function embedQuery(searchQuery: string): Promise<number[]> {
  const [embedding] = await embedTexts([searchQuery], "retrieval.query");
  return embedding;
}
